### PR TITLE
Change choice handling to consider both depends on and visibility constraint

### PR DESCRIPTION
**Pull request by jeho-oh**
Choice and choice values are selectable only when both their depends on and visibility constraints are satisfied.
Regarding this, the following changes are made:
- Condition for choice now uses correct visibility condition and does not use dir_dep and rev_dep.
- Kclause uses visibility of choice values (which has depends on constraint conjuncted by kconfig) to generate choice clauses.

In addition, there is a small optimization to avoid generating redundant clauses for the choice mutex.

### CODE REVIEW

File: kmax/kclause [hunk 1]
```diff
@@ -607,11 +607,11 @@ if __name__ == '__main__':
     # TODO: Update choice_vars with tristate_choice as well when print_dimacs() method
     # is modified for tristate_choices.
     choice_vars.update(set(config_vars))
-    for i in range(0, len(config_vars)):
+    for i in range(0, len(config_vars)-1):
       defined_vars.add(config_vars[i])
```
Comment (necipfazil) [inline kmax/kclause:611]: Here, with this change, we no longer add the last element of config_vars to defined_vars set. Is this intended?

Comment (necipfazil) [reply kmax/kclause:611]: Later in the code we add all `config_vars` to `defined_vars` while _generating 'at least one' clause for boolean and tristate choices_. So, it seems like we can remove this set update.

Comment (jeho-oh) [reply kmax/kclause:611]: Good point. I'll remove line 611 on the next commit.

**Comment by necipfazil** [comment]
Thanks for fixing several points in the code. The changes LGTM.

I pushed a few commits that removes the code commented out as they are replaced with their fixed versions -- and did some refactoring.

**Review by necipfazil — APPROVED** [review]

**Review by jeho-oh — COMMENTED** [review]
Looks good. Thanks!

**Pull request merged**